### PR TITLE
DSDF Test-Time Feature Alignment for OOD Tandem Generalization

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The model's DSDF feature distribution for NACA6416 (OOD, heavily cambered) systematically differs from the training distribution (mostly NACA00XX symmetric foils). When the learned mapping `f(dsdf_features) → flow field` is evaluated on inputs from a low-density region of its training distribution, prediction accuracy degrades — this is the standard covariate shift problem, and it directly contributes to the p_tan gap (28.50 vs p_in 13.21, a 2.17x ratio).

**Key insight:** This is not a model capacity problem — it is an input distribution mismatch problem. If we align the per-sample DSDF channel statistics of OOD test samples to the training distribution before inference, we pull the model's inputs back into its trained domain without any change to model weights or training procedure.

This is the inference-time version of Domain Adaptation. Similar to test-time batch norm adaptation (Schneider et al., NeurIPS 2020) but applied to raw DSDF input features rather than batch norm statistics.

**Why this is different from SWD Domain Alignment (#2175, CLOSED):** SWD aligned internal slice token distributions during training via a loss — destroyed physically meaningful tandem-specific representations. This approach aligns raw input features at test time only (no training change) and only for OOD tandem samples. Cannot harm training-domain accuracy.

**Literature:**
- Schneider et al. (2020). "Improving robustness against common corruptions by covariate shift adaptation." NeurIPS 2020. Test-time BN adaptation → 5-10% improvement on distribution-shifted inputs.
- Wang et al. (2021). "Tent: Fully Test-Time Adaptation by Entropy Minimization." ICLR 2021. Canonical TTA paper.
- Sun & Saenko (2016). "Deep CORAL: Correlation Alignment for Deep Domain Adaptation." Second-order statistics alignment; we use simpler first-order only.

### Step 1: Compute training DSDF channel statistics

Compute per-channel mean and std for DSDF channels (x[:, :, 2:10], 8 channels) across all training samples. Hardcode as constants in train.py.

```python
# One-time scan of training set:
all_dsdf = []
for batch in train_dataloader:
    x = batch['x']  # (B, N, C)
    all_dsdf.append(x[:, :, 2:10].reshape(-1, 8))
all_dsdf = torch.cat(all_dsdf, dim=0)  # (N_total_nodes, 8)
train_dsdf_mean = all_dsdf.mean(dim=0)  # (8,)
train_dsdf_std = all_dsdf.std(dim=0)    # (8,)
# Print and hardcode these as DSDF_TRAIN_MEAN, DSDF_TRAIN_STD constants
```

Or compute lazily on first validation pass and cache.

### Step 2: Apply alignment during validation only

```python
def apply_dsdf_tta(x, train_dsdf_mean, train_dsdf_std, tandem_mask):
    """
    Align DSDF feature distribution of OOD samples to training distribution.
    Apply ONLY during validation, ONLY to tandem OOD samples.
    
    x: (B, N, C) full feature tensor
    tandem_mask: (B,) bool — True if this sample is an OOD tandem sample
    train_dsdf_mean: (8,) training distribution mean per channel
    train_dsdf_std: (8,) training distribution std per channel
    """
    if tandem_mask.sum() == 0:
        return x
    
    x = x.clone()
    dsdf_slice = slice(2, 10)
    train_dsdf_mean = train_dsdf_mean.to(x.device)
    train_dsdf_std = train_dsdf_std.to(x.device)
    
    for b in range(x.shape[0]):
        if tandem_mask[b]:
            dsdf = x[b, :, dsdf_slice]  # (N, 8)
            sample_mean = dsdf.mean(dim=0)           # (8,)
            sample_std = dsdf.std(dim=0).clamp(min=1e-6)  # (8,)
            dsdf_aligned = (dsdf - sample_mean) / sample_std * train_dsdf_std + train_dsdf_mean
            x[b, :, dsdf_slice] = dsdf_aligned
    return x
```

In the validation loop:
```python
if args.dsdf_tta_align and not model.training:
    x = apply_dsdf_tta(x, DSDF_TRAIN_MEAN, DSDF_TRAIN_STD, tandem_mask)
```

### Step 3: Flag

- `--dsdf_tta_align`: Enable DSDF test-time feature alignment (validation only)

### Step 4: Run experiments

```bash
# Seed 42
cd cfd_tandemfoil && python train.py --agent tanjiro --seed 42 \
  --wandb_name "tanjiro/dsdf-tta-align-s42" \
  --wandb_group "tanjiro/dsdf-tta-align" \
  --dsdf_tta_align \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5

# Seed 73: same command, --seed 73, --wandb_name "tanjiro/dsdf-tta-align-s73"
```

### Key implementation notes

1. **Identifying OOD tandem samples:** Check how the dataset labels tandem vs single-foil samples. The OOD tandem val split (p_tan) contains NACA6416 samples. Use whatever flag/mask the dataset provides to identify these samples during validation.

2. **DSDF channel indices:** Verify channels 2:10 are the 8 DSDF channels. Print feature stats for one sample to confirm by magnitude range.

3. **Apply ONLY at validation:** The training pipeline uses DSDF magnitude aug (aug_dsdf2_sigma) which deliberately shifts training distributions. The TTA alignment must ONLY touch the validation forward pass.

4. **Sanity check:** Log pre/post alignment mean/std for one OOD sample in the first validation epoch. Should converge to training distribution values.

5. **Variant to try if Variant A fails:** Apply alignment to fore-foil DSDF (2:6) and aft-foil DSDF (6:10) separately — they may have different training distributions. Try separate alignment if global alignment fails.

## Baseline

| Metric | 2-seed avg | Target to beat |
|--------|-----------|----------------|
| p_in | **13.21** | < 13.21 |
| p_oodc | **7.82** | < 7.82 |
| **p_tan** | **28.50** | **< 28.50** |
| p_re | **6.45** | < 6.45 |

**Baseline W&B runs:** 6yfv5lio (s42, p_tan=28.432), etepxvjc (s73, p_tan=28.572) — PR #2184 (DCT Freq Loss w=0.05)

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent tanjiro --wandb_name "tanjiro/baseline-dct-freq" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5
```

### CODE REVIEW

**Comment by tcapelle** [comment]
## Experiment Results — DSDF Test-Time Feature Alignment

### Implementation
- Computed per-channel mean/std for DSDF channels (x[:,:,2:10], 8 channels) across all training samples
- At validation time only, for tandem samples only: standardized per-sample DSDF → rescaled to training distribution stats
- `--dsdf_tta_align` flag, applied before x-standardization in val loop

### DSDF Training Channel Stats
- Mean: [-0.008, 0.142, 3.688, 3.431, 3.439, 3.587, 3.780, 3.429]
- Std: [1.070, 1.279, 1.442, 2.062, 2.121, 2.055, 1.324, 2.097]

### Results (2 seeds)

| Metric | Baseline (2-seed avg) | TTA s42 (3wr2j3e2) | TTA s73 (kwpqw6gv) | TTA avg | Delta |
|--------|----------------------|---------------------|---------------------|---------|-------|
| p_in | **13.21** | 12.7 | 13.4 | 13.05 | -1.2% |
| p_oodc | **7.82** | 31.2 | 39.8 | 35.50 | +354% |
| **p_tan** | **28.50** | 47.5 | 48.9 | **48.20** | **+69%** |
| p_re | **6.45** | 25.3 | 33.0 | 29.15 | +352% |

Both ran to epoch 149 (180 min timeout, ~72s/epoch).

### Analysis

**The TTA alignment catastrophically degrades all OOD metrics.** p_tan went from 28.50 → 48.20 (+69%), p_oodc from 7.82 → 35.50 (+354%), p_re from 6.45 → 29.15 (+352%). Only p_in showed a marginal improvement.

**Why it failed:** The alignment replaces each sample's per-channel DSDF mean/std with the training distribution's global mean/std. This destroys the geometry-specific information encoded in the DSDF features — the model uses absolute DSDF values to understand foil shape and proximity, not just relative patterns. Normalizing away the sample-specific statistics removes exactly the signal the model needs to differentiate geometries.

The x-standardization (line 2350: `x = (x - stats["x_mean"]) / stats["x_std"]`) already handles bringing features into a consistent range. The TTA alignment effectively double-normalizes the DSDF channels, collapsing the geometry signal.

**This approach is a clear dead end.** The covariate shift hypothesis may have some validity for p_tan, but the intervention needs to be more targeted — perhaps aligning only specific higher-order statistics or working in the model's latent space rather than raw input space.

### Suggested follow-ups
1. **Latent-space TTA**: Adapt batch norm or layer norm statistics inside DomainLayerNorm at test time instead of raw inputs
2. **Partial alignment**: Only align the 4 DSDF channels (x[:,:,4:8] or x[:,:,6:10]) rather than all 8 including SAF
3. **Focus on p_tan through other means**: The tandem gap is more likely an architecture/loss problem than an input distribution problem

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2189 because ADVISOR: Closing — clear dead end. DSDF TTA alignment catastrophically degraded all OOD metrics (p_tan +69%, p_oodc +354%, p_re +352%). The per-sample normalization destroyed the geometry-specific information encoded in absolute DSDF values — the model uses these to differentiate foil shapes and proximity. Combined with SWD (#2175) and MixStyle (#2188), this confirms that raw-input and feature-space distribution manipulation is counterproductive for this problem: the 'covariate shift' in DSDF features IS the geometry signal.

Your analysis was excellent — especially the insight about double-normalization with x-standardization. The student suggestion about latent-space TTA (adapting DomainLayerNorm statistics) is worth noting for future reference, though output/decoder-side approaches are more promising.

Well done on the thorough investigation. Moving you to a new experiment..
